docs/arch/adr-03.md

```diff
@@ -0,0 +1,115 @@
+
+ADR 3: Position sampling and patches of the sky
+===============================================
+
+Author: Nicolas Tessore  
+Date: 5 February 2021  
+Status: draft
+
+
+Context
+-------
+
+This ADR addresses two related open problems for SkyPy simulations:
+
+- How can we describe the regions in which to sample the positions of e.g.
+  galaxies or supernovae?
+- How can we break up the sampling over large fractions of the sky into smaller
+  chunks?
+
+The second point is particularly relevant when the simulation becomes so large
+that it no longer fits into memory, as well as for parallelisation.
+
+
+Extension to pipelines
+----------------------
+
+This ADR proposes to introduce a new top-level keyword `regions` [alt: `sky`,
+`geometry`, `patches`] into pipelines that describes the geometry of the
+simulation. For example, to include a "rectangular" and a "circular" region:
+
+```yaml
+regions:
+- !rectangular [ ... ]
+- !circular [ ... ]
+```
+
+The `regions` list can be traversed by the pipeline runner to create what are
+effectively independent parallel simulations. The list items are objects with
+the following interface.
+
+
+Region interface
+----------------
+
+The regions need to support two sets of operations:
+
+- Information queries: For example, there should be a `.area` [alt:
+  `.solid_angle`] attribute that returns the solid angle of the region.
+- Random sampling: There needs to be at least a `random_point()` [alt:
+  `random_position()`, `uniform_in()`] function that can uniformly sample a
+  random position from within the region.
+
+When the pipeline runner traverses the list of regions, it can keep track of
+the current region in a `$region` reference that can be used where necessary.
+For example, to sample from a luminosity function with positions:
+
+```yaml
+tables:
+  galaxies:
+    z, M: !schechter_lf
+      ...
+      sky_area: $region.area
+    ra, dec: !random_point [ $region ]
```

Comment:
I am a bit out of date with how SkyPy config files work, is this the same `yaml` file as above or a separate file?

Reply:
This is the same file. The top-level region specification is looped over in the pipeline class, and each loop creates essentially a table fragment as specified here.